File: BinarySearch/1712. Ways to Split Array Into Three Subarrays/1712. Ways to Split Array Into Three Subarrays.java
class Solution {
    public int waysToSplit(int[] nums) {
        long mod = 1000000007;
        int n = nums.length;
        long[] presum = new long[n];
        presum[0] = nums[0];
        for (int i = 1; i <n; i++) {
            presum[i] = presum[i-1] + nums[i];
        }
        
        long rets = 0;
        //long total = presum[n];
        
        int j = 0, k = 0;
        for (int i = 0; i < n; i++) { // j, k not recover to initial postion, it can continue to increasing their pos
            while (j<= i || (j < n && presum[j]-presum[i] <presum[i]) ) j++;
            if (j == n) break;
            
            while (k+1 < n-1 && 2*presum[k+1]<=presum[n-1]+presum[i]) k+=1;
            if (k < j) continue;
            
            rets += k-j+1;
            
            rets %= mod;
        }
        return (int)rets;
    }
}

// [X X X] [X X X| X X X] X X X
//      i        j     k      n-1
// k is a upper bound for the second subarray
// presum[k]-presum[i]<=presum[n]-presum[k]
// 2*presum[k]<=presum[n]+presum[i]

File: BinarySearch/1712. Ways to Split Array Into Three Subarrays/1712. Ways to Split Array Into Three Subarrays.cpp
class Solution {
public:
    int waysToSplit(vector<int>& nums) {
        int n = nums.size();
        vector<long long>presum(n, 0);
        presum[0] = nums[0];
        for (int i = 1; i < n; i++) {
            presum[i] = presum[i-1] + nums[i];
        }
        
        int j = 0;
        long long rets = 0;
        long long M = 1e9+7;
        for (int i = 0; i < n ; i++) {
            while (j <= i || (j<n && presum[j]-presum[i] < presum[i]) ) {
                j++;
            }
            if (j == n) break;
            
            auto iter = upper_bound(presum.begin(), presum.end(), 0.5*(presum[i] + presum[n-1])) - 1;
            
            
            int k = iter-presum.begin();
            //cout<<(*iter)<<" "<<j<<" "<<(*presum.begin())<<" "<<k<<endl;
            if (k < j) continue;
            if (k == n-1) k--;
            
            rets += k-j+1;
            rets %= M;
        }
        return rets;
    }
};
